File: rust/src/linked_lists/singly.rs
use core::ptr::NonNull;

use super::node::SinglyNode as Node;

pub struct SinglyList<T> {
    head: Box<Node<T>>,
    tail: NonNull<Node<T>>,
    size: usize,
}

impl<T: Default> SinglyList<T> {
    pub fn new() -> Self {
        let mut sentinel = Box::new(Node::new(T::default(), None));
        let sentinel_ptr = NonNull::from(sentinel.as_mut());

        SinglyList {
            head: sentinel,
            tail: sentinel_ptr,
            size: 0,
        }
    }

    pub fn len(&self) -> usize {
        self.size
    }

    pub fn is_empty(&self) -> bool {
        self.size == 0
    }

    pub fn push_back(&mut self, value: T) {
        let mut new_node = Box::new(Node::new(value, None));
        let new_tail = NonNull::from(new_node.as_mut());

        unsafe {
            self.tail.as_mut().next = Some(new_node);
        }

        self.tail = new_tail;
        self.size += 1;
    }

    pub fn push_front(&mut self, value: T) {
        let mut new_node = Box::new(Node::new(value, self.head.next.take()));
        let new_node_ptr = NonNull::from(new_node.as_mut());

        self.head.next = Some(new_node);
        if self.size == 0 {
            self.tail = new_node_ptr;
        }

        self.size += 1;
    }

    pub fn insert_at(&mut self, pos: usize, value: T) {
        if pos > self.size {
            panic!("index out of bounds");
        }

        if pos == 0 {
            self.push_front(value);
        } else if pos == self.size {
            self.push_back(value);
        } else {
            let mut curr = self.head.as_mut();
            for _ in 0..pos {
                curr = curr.next.as_mut().unwrap().as_mut();
            }

            let new_node = Box::new(Node::new(value, curr.next.take()));
            curr.next = Some(new_node);

            self.size += 1;
        }
    }

    pub fn pop_back(&mut self) -> Option<T> {
        match self.size {
            0 | 1 => self.pop_front(),
            _ => {
                let mut curr = self.head.as_mut();
                for _ in 0..self.size - 1 {
                    curr = curr.next.as_mut().unwrap().as_mut();
                }

                let deleted_node = curr.next.take().unwrap();

                self.tail = NonNull::from(curr);

                self.size -= 1;

                Some(deleted_node.value)
            }
        }
    }

    pub fn pop_front(&mut self) -> Option<T> {
        if self.size == 0 {
            return None;
        }

        let mut first_node = self.head.next.take().unwrap();
        self.head.next = first_node.next.take();

        self.size -= 1;

        if self.len() == 0 {
            self.tail = NonNull::from(self.head.as_mut());
        }

        Some(first_node.value)
    }

    pub fn delete_at(&mut self, pos: usize) -> Option<T> {
        if pos >= self.size {
            panic!("index out of bounds");
        }

        if pos == 0 {
            self.pop_front()
        } else if pos == self.size - 1 {
            self.pop_back()
        } else {
            let mut curr = self.head.as_mut();
            for _ in 0..pos {
                curr = curr.next.as_mut().unwrap().as_mut();
            }

            let node_to_delete = curr.next.take().unwrap();
            curr.next = Some(node_to_delete.next.unwrap());

            self.size -= 1;

            Some(node_to_delete.value)
        }
    }

    pub fn update_at(&mut self, pos: usize, value: T) {
        if pos >= self.size {
            panic!("index out of bounds");
        }

        if pos == self.size - 1 {
            unsafe {
                self.tail.as_mut().value = value;
            }
        } else {
            let mut curr = self.head.as_mut();
            for _ in 0..pos {
                curr = curr.next.as_mut().unwrap().as_mut();
            }

            curr.value = value;
        }
    }

    pub fn peek_back(&self) -> Option<&T> {
        if self.size == 0 {
            return None;
        }
        unsafe { Some(&self.tail.as_ref().value) }
    }

    pub fn peek_back_mut(&mut self) -> Option<&mut T> {
        if self.size == 0 {
            return None;
        }
        unsafe { Some(&mut self.tail.as_mut().value) }
    }

    pub fn peek_front(&self) -> Option<&T> {
        if self.size == 0 {
            return None;
        }
        Some(&self.head.next.as_ref().unwrap().value)
    }

    pub fn peek_front_mut(&mut self) -> Option<&mut T> {
        if self.size == 0 {
            return None;
        }
        Some(&mut self.head.next.as_mut().unwrap().value)
    }

    pub fn get(&self, pos: usize) -> Option<&T> {
        if pos >= self.size {
            panic!("index out of bounds");
        }

        if pos == 0 {
            self.peek_back()
        } else if pos == self.size - 1 {
            self.peek_front()
        } else {
            let mut curr = self.head.as_ref();
            for _ in 0..pos {
                curr = curr.next.as_ref().unwrap().as_ref();
            }

            Some(&curr.value)
        }
    }

    pub fn get_mut(&mut self, pos: usize) -> Option<&mut T> {
        if pos >= self.size {
            panic!("index out of bounds");
        }

        if pos == 0 {
            self.peek_back_mut()
        } else if pos == self.size - 1 {
            self.peek_front_mut()
        } else {
            let mut curr = self.head.as_mut();
            for _ in 0..pos {
                curr = curr.next.as_mut().unwrap().as_mut();
            }

            Some(&mut curr.value)
        }
    }

    pub fn clear(&mut self) {
        while self.size != 0 {
            self.pop_front();
        }
    }
}

File: c/src/linked_lists/singly.c
#include "linked_lists/singly.h"

#include <stdlib.h>

Node* node_new(void* value, Node* next) {
    Node* new_node = malloc(sizeof(Node));
    new_node->value_ = value;
    new_node->next_ = next;
    return new_node;
}


SinglyLinkedList* singly_list_new() {
    SinglyLinkedList* new_list = malloc(sizeof(SinglyLinkedList));

    Node* sentinel = node_new(NULL, NULL);

    new_list->sentinel_ = sentinel;
    new_list->tail_ = sentinel;
    new_list->size_ = 0;

    return new_list;
}

size_t size(SinglyLinkedList* list) {
    return list->size_;
}

bool is_empty(SinglyLinkedList* list) {
    return list->size_ == 0;
}

void* front(SinglyLinkedList* list) {
    return list->size_ > 0 ? list->sentinel_->next_->value_ : NULL;
}

void* back(SinglyLinkedList* list) {
    return list->size_ > 0 ? list->tail_->value_ : NULL;
}
    
void* at(SinglyLinkedList* list, size_t pos) {
    if (pos < 0 || pos >= list->size_) return NULL;

    if (pos == 0) return front(list);
    else if (pos == list->size_-1) return back(list);
    else {
        Node* curr = list->sentinel_;
        while (pos--) curr = curr->next_;
        return curr->value_;
    }
}

Node* search(SinglyLinkedList* list, void* value) {
    Node* curr = list->sentinel_->next_;
    while (curr) {
        if (curr->value_ == value) return curr;
    }
    return NULL;
}

void push_front(SinglyLinkedList* list, void* value) {
    Node* new_node = node_new(value, list->sentinel_->next_);
    list->sentinel_->next_ = new_node;

    if (list->size_ == 0) list->tail_ = new_node;

    list->size_++;
}

void push_back(SinglyLinkedList* list, void* value) {
    Node *new_node = node_new(value, NULL);
    
    if (list->tail_) list->tail_->next_ = new_node;
    else list->sentinel_->next_ = new_node;
    list->tail_ = new_node;

    list->size_++;
}

void insert(SinglyLinkedList* list, size_t pos, void* value) {
    if (pos < 0 || pos > list->size_) return;

    if (pos == 0) push_front(list, value);
    else if (pos == list->size_) push_back(list, value);
    else {
        Node* curr = list->sentinel_;
        while (pos--) curr = curr->next_;

        Node* new_node = node_new(value, curr->next_);
        curr->next_ = new_node;

        list->size_++;
    }
}

void* pop_front(SinglyLinkedList* list) {
    if (list->size_ == 0) return NULL;

    Node* prev_head = list->sentinel_->next_;
    void* prev_head_value = prev_head->value_;
    list->sentinel_->next_ = prev_head->next_;
    free(prev_head);
    if (list->size_ == 1) list->tail_ = list->sentinel_;

    list->size_--;

    return prev_head_value;
}

void* pop_back(SinglyLinkedList* list) {
    if (list->size_ == 0) return NULL;

    Node* curr = list->sentinel_;
    while (curr->next_ != list->tail_) curr = curr->next_;

    Node* prev_tail = list->tail_;
    void* prev_tail_value = prev_tail->value_;
    free(prev_tail);
    list->tail_ = curr;
    list->tail_->next_ = NULL;

    list->size_--;

    return prev_tail_value;
}

void* erase(SinglyLinkedList* list, size_t pos) {
    if (pos < 0 || pos > list->size_) return NULL;

    if (pos == 0) return pop_front(list);
    else if (pos == list->size_-1) return pop_back(list);
    else {
        Node* curr = list->sentinel_;
        while (curr--) curr = curr->next_;

        Node* node_to_delete = curr->next_;
        void* deleted_node_value = node_to_delete->value_;
        curr->next_ = node_to_delete->next_;
        free(node_to_delete);

        list->size_--;

        return deleted_node_value;
    }
}

void clear(SinglyLinkedList* list) {
    while (list->size_ != 0) pop_front(list);
}

Node* find_middle(Node* head) {
    if (!head) return NULL;

    Node* slow = head;
    Node* fast = head->next_;

    while (fast && fast->next_) {
        slow = slow->next_;
        fast = fast->next_->next_;
    }

    return slow;
}

Node* merge_helper(Node* head_1, Node* head_2) {
    Node* dummy = node_new(NULL, NULL);
    Node* curr = dummy;

    while (head_1 && head_2) {
        if (head_1->value_ <= head_2->value_) {
            curr->next_ = head_1;
            head_1 = head_1->next_;
        } else {
            curr->next_ = head_2;
            head_2 = head_2->next_;
        }

        curr = curr->next_;
    }

    Node* sorted_head = dummy->next_;
    free(dummy);

    return sorted_head;
}

Node* sort_list(Node* head) {
    if (!head || !head->next_) return head;

    Node* middle = find_middle(head);
    Node* rigth_head = middle->next_;
    middle->next_ = NULL;

    Node* sorted_left_head = sort_list(head);
    Node* sorted_right_head = sort_list(rigth_head);

    return merge_helper(sorted_left_head, sorted_right_head);
}

void sort(SinglyLinkedList* list) {
    list->sentinel_ = sort_list(list->sentinel_->next_);
}

SinglyLinkedList* merge(SinglyLinkedList* list_1, SinglyLinkedList* list_2) {
    if (!is_empty(list_1) && is_empty(list_2)) return list_1;
    if (is_empty(list_1) && !is_empty(list_2)) return list_2;

    SinglyLinkedList* new_list = malloc(sizeof(SinglyLinkedList));
    new_list->sentinel_ = list_1->sentinel_;
    new_list->tail_ = list_2->tail_;
    new_list->size_ = list_1->size_ + list_2->size_;

    list_1->tail_->next_ = list_2->sentinel_->next_;
    free(list_2->sentinel_);

    return new_list;
}
